lib/l10n_utils/fluent.py

```diff
@@ -0,0 +1,87 @@
+from functools import wraps
+from hashlib import sha256
+
+from django.conf import settings
+from django.core.cache import caches
+from django.utils.encoding import force_bytes
+
+from fluent.runtime import FluentLocalization, RootedFileResourceLoader
+
+from lib.l10n_utils import translation
+
+
+cache = caches['l10n']
+
+
+def _cache_key(*args, **kwargs):
+    return sha256(force_bytes(f'fluent:{args}:{kwargs}')).hexdigest()
+
+
+def memoize(f):
+    """Decorator to cache the results of expensive functions"""
+    @wraps(f)
+    def inner(*args, **kwargs):
+        key = _cache_key(f.__name__, *args, **kwargs)
+        value = cache.get(key)
+        if value is None:
+            value = f(*args, **kwargs)
+            cache.set(key, value)
+
+        return value
+
+    return inner
+
+
+@memoize
+def fluent_bundle(locales, files):
+    if isinstance(locales, str):
+        locales = [locales]
+
+    # file IDs may not have file extension
+    files = [f'{f}.ftl' for f in files if not f.endswith('.ftl')]
+    # temporary until MultiRootLoader lands
+    path = f'{settings.FLUENT_PATHS[1]}/{{locale}}/'
+    loader = RootedFileResourceLoader(path)
+    return FluentLocalization(locales, files, loader)
+
+
+def has_message(message_id, bundle):
+    # assume English locales have the message
+    if bundle.locales[0].startswith('en-'):
+        return True
+
+    if not bundle._bundle_cache:
+        # need to warm the cache in the bundle
+        bundle.format_value('warm')
```

Comment:
This one is worst-case performance, it will load the full fallback chain always.

I'd be up for taking this API into python-fluent, too.

Big picture question, though:

This is part of the activation strategy? Should that be a runtime thing or a deployment thing?

Reply:
I've reworked this a good bit. The use case is that in the template we need to be able to know whether a translation has a message_id or not so that we can show some new part of the page or keep it hidden until the translation of the new string(s) is done.